### PR TITLE
NewtonIterationInterleaved: allow np to be different than 3.

### PR DESCRIPTION
**Pull request by dr-robertk**
This PR changes all places in the NewtonIterationInterleaved such that np could be different than 3, as for example needed for flow_solvent. For now, np is set to 3 and the code works as before.

However, np should then be a template parameter and I leave this change for @atgeirr because it involves some redesign of the BlackoilModel.

### CODE REVIEW

**Comment by bska** [comment]
Cool stuff.  I have one question (genuine, I don't know the answer).  Would it make sense to reverse to order of the loops in
```C++
const int size = istlA.N();
Vector istlb(size);
for (int i = 0; i < size; ++i) {
    for (int phase = 0, idx = i; phase<np; ++phase, idx += size) {
        istlb[i][ phase ] = b(idx );
    }
}
```
so that it becomes
```C++
int idx = 0;
for (int phase = 0; phase < np; ++phase) {
    for (int i = 0; i < size; ++i, ++idx) {
        istlb[i][phase] = b(idx);
    }
}
```
?  That way we get unit stride into `b` and `np` stride into `istlb` rather than unit stride into `istlb` and `size` stride into `b`.  Also, the inner loop ranges over the cells rather than the phases which may help the instruction scheduler on the CPU.  I would like to know whether or not that makes any measurable difference at all.

As for making `np` a template parameter, that's a design discussion.  We may want to experiment with dynamically sized blocks (i.e., `Dune::DenseMatrix` and `Dune::DenseVector`) rather than the current, fixed-size scheme.

File: opm/autodiff/NewtonIterationBlackoilInterleaved.cpp [hunk 1]
```diff
@@ -234,8 +238,6 @@ namespace Opm
         assert(size == row_major.cols());
 
         // Create ISTL matrix with interleaved rows and columns (block structured).
-        assert(np == 3);
-        istlA.setSize(row_major.rows(), row_major.cols(), row_major.nonZeros());
```
Comment (atgeirr) [inline opm/autodiff/NewtonIterationBlackoilInterleaved.cpp]: Are you sure that this is not necessary, or beneficial?

**Comment by atgeirr** [comment]
This is a good start, and I agree that we should make this a template. I am ok with merging this branch now as a starting point, or we can work on this branch together and hold off merging till we're done. What do you prefer?

**Comment by dr-robertk** [comment]
I will look into this the next days. After the latest merges I was this code was not running anymore. 
I will revise to incorporate this into DuneMatrix.

**Comment by atgeirr** [comment]
> I will revise to incorporate this into DuneMatrix.

I am not sure what you intend to do?

**Comment by dr-robertk** [comment]
I want to put the formInterleavedSystem into the DuneMatrix class. One might want to uses it elsewhere.

**Comment by atgeirr** [comment]
> I want to put the formInterleavedSystem into the DuneMatrix class. One might want to uses it elsewhere.

I see. I also want to use the formInterleavedSystem() function (or a relative) to use in different cases, too, so I'll be watching this with interest. This would mean making DuneMatrix a template, right?

**Comment by bska** [comment]
> I want to put the `formInterleavedSystem()` into the `DuneMatrix` class

What are the benefits to making it a member function versus a simple, free function?  Right now, `formInterleavedSystem()` does not access any data members of the enclosing class and can be implemented entirely in terms of public APIs.

**Comment by dr-robertk** [comment]
I think that like in the CPR preconditioner it is a good idea to simply write DuneMatrix( Eigen::SparseMatrix ). When np=1 (of DuneMatrix) is one we do what we did before, otherwise you have to store in the interleaved mode anyway. That's why I think it's a good idea to have this in once place.

**Comment by dr-robertk** [comment]
This has been replaced by #496 .

**Pull request closed without merging**